datajunction-server/datajunction_server/construction/build_v2.py

```diff
@@ -0,0 +1,938 @@
+                dimensional_columns.append(expr)
+                expr.set_semantic_type(SemanticType.DIMENSION)  # type: ignore
+            else:
+                measure_columns.append(expr)
+                expr.set_semantic_type(SemanticType.MEASURE)  # type: ignore
+        await parent_ast.compile(context)
+
+        # Build translated SQL object
+        columns_metadata = [
+            assemble_column_metadata(  # pragma: no cover
+                cast(ast.Column, col),
+            )
+            for col in parent_ast.select.projection
+        ]
+        dependencies, _ = await parent_ast.extract_dependencies(
+            CompileContext(session, DJException()),
+        )
+        measures_sql_mapping[parent_node.name] = TranslatedSQL(
+            sql=str(parent_ast),
+            columns=columns_metadata,
+            dialect=build_criteria.dialect,
+            upstream_tables=[
+                f"{dep.catalog.name}.{dep.schema_}.{dep.table}"
+                for dep in dependencies
+                if dep.type == NodeType.SOURCE
+            ],
+        )
+    return measures_sql_mapping
+
+
+async def build_node(  # pylint: disable=too-many-arguments,too-many-locals
+    session: AsyncSession,
+    node: NodeRevision,
+    filters: Optional[List[str]] = None,
+    dimensions: Optional[List[str]] = None,
+    orderby: Optional[List[str]] = None,  # pylint: disable=unused-argument
+    limit: Optional[int] = None,  # pylint: disable=unused-argument
+    build_criteria: Optional[BuildCriteria] = None,
+    access_control: Optional[access.AccessControlStore] = None,
+    include_dimensions_in_groupby: bool = None,
+) -> ast.Query:
+    """
+    Builds the node SQL with the requested set of dimensions, filter expressions,
+    order by, and limit clauses.
+
+    Build Strategy
+    ---------------
+    1. Recursively turn node references into query ASTs + apply any filters that can
+    be pushed down. If the node query has CTEs, unwind them into the query.
+    2. Initialize the final query with the node's query AST added to it as a CTE.
+    3. For any dimensions or filters requested for the node, determine if a join is
+    needed to bring in the dimension or filter. Keep track of all the necessary dimension
+    joins in a dict that maps dimension nodes to join paths.
+    4. For each of the necessary dimension joins, build the dimension node's query in the
+    same manner as above, recursively replacing any node references and pushing down requested
+    filters where possible.
+    5. Add each dimension node's query AST to the final query as a CTE.
+    6. Build the final query using the various CTEs. This does all the joins between the node
+    query AST and the dimension nodes' ASTs using the join logic from the dimension links.
+    7. Add all requested dimensions to the final select.
+    8. Add order by and limit to the final select (TODO)
```

Comment:
Looking at this build strategy has me thinking--should we use the [builder pattern](https://refactoring.guru/design-patterns/builder/python/example) here since there are a series of complicated steps we need to do in order? A few benefits I can think of is:
- It will make the order of operations easier to understand by just looking at the method chain that happens in the build
- It will let us keep a shared context for the build instance where we can store warnings/errors
- When a particular section of the build fails, the error at the top of the chain can explicitly say that, like `DJBuildException("Failed to build dimension queries")

Reply:
Great idea, this pattern would help! I'll do another round of refactoring now that we have some basic tests in place.

Reply:
Awesome! The example in that link does a bit more than I think we'd need to do here like create an abstract base class and stuff. I think a single builder class that carries all of these step functions and a `build` method that calls them in the right order (and catches and re-raises errors for each step) is good enough. Then the return type can just be an immutable fully built query.

Reply:
Oh yeah, we don't need that. I was just going to add a `QueryBuilder` class that stores errors and allows for building each of the arguments to this function separately (i.e., `with_orderby`, or `with_limit`, neither of which I've implemented at the moment).